swim/memberlist.go

```diff
@@ -320,50 +387,52 @@ func (m *memberlist) Update(changes []Change) (applied []Change) {
 	m.node.emit(MemberlistChangesReceivedEvent{changes})
 
 	m.Lock()
-	m.members.Lock()
 
+	// run through all changes received and figure out if they need to be accepted
+	m.members.Lock()
 	for _, change := range changes {
-		member, ok := m.members.byAddress[change.Address]
+		member, has := m.members.byAddress[change.Address]
+
+		// transform the change into a member that we can test against existing
+		// members
+		gossip := Member{}
+		gossip.populateFromChange(&change)
+
+		// test to see if we accpet the gossip
+		if acceptGossip(member, &gossip) {
+			// the gossip is accepted
+
+			if gossip.Address == m.local.Address {
+				// if the gossip is about the local member it needs to be
+				// countered with a reincarnation gossip
+				change = m.reincarnationChange()
+				m.node.emit(RefuteUpdateEvent{})
+			} else {
+				// otherwise it can be applied to the memberlist
+
+				// if the member was not already present in the list we will it
+				// needs to be added and randomly inserted in the list to ensure
+				// guarantees for pinging
+				if !has {
+					m.members.byAddress[gossip.Address] = &gossip
+					i := m.getJoinPosition()
+					m.members.list = append(m.members.list[:i], append([]*Member{&gossip}, m.members.list[i:]...)...)
+				} else {
+					// copy the state of the gossip to the member
+					*member = gossip
```

Comment:
It might improve understandability if you change this to `m.members.byAddress[gossip.Address] = &gossip` or make it more clear with the comment that that's what you're basically doing.